store/prefixstore.go

```diff
@@ -124,3 +168,43 @@ func (iter prefixIterator) Value() []byte {
 func (iter prefixIterator) Close() {
 	iter.iter.Close()
 }
+
+// copied from github.com/tendermint/tendermint/libs/db/prefix_db.go
+func stripPrefix(key []byte, prefix []byte) []byte {
+	if len(key) < len(prefix) || !bytes.Equal(key[:len(prefix)], prefix) {
+		panic("should not happen")
+	}
+	return key[len(prefix):]
+}
+
+// wrapping sdk.PrefixEndBytes
+func cpIncr(bz []byte) []byte {
+	return sdk.PrefixEndBytes(bz)
+}
+
+// copied from github.com/tendermint/tendermint/libs/db/util.go
+func cpDecr(bz []byte) (ret []byte) {
+	if len(bz) == 0 {
+		panic("cpDecr expects non-zero bz length")
+	}
+	ret = clone(bz)
+	for i := len(bz) - 1; i >= 0; i-- {
+		if ret[i] > byte(0x00) {
+			ret[i]--
+			return
+		}
+		ret[i] = byte(0xFF)
```

Comment:
Actually this is okay in the context of where it should be used in `ReverseIterator` 

```
0x23 0x00 0x00 <- prefix
0x22 0xff 0xff <- current implementation of cpDecr (skipped by iterator)
...
0x22 <- cpDecr output consistent with new cpIncr (skipped by iterator)
```

I still think it would be more consistent to have the same truncation logic for both. What would be better is if we could just have inclusive end points since `0x23` is the tight lower bound for the prefix `0x22 0xff 0xff`, just like `0x23 0x00 0x01` is the tight upper bound

Reply:
0x22 I think would cause iteration over too many trailing items, like 0x22 0x01 ... 

also, I'm not convinced that inclusive endpoints would be good...  it seems asymmetric w/ forward Iterator.

Reply:
Yeah we always have that problem though, e.g. `0x22 0xff 0xff 0x01` > `0x22 0xff 0xff`. That's why we have the skip logic in the iterator here: https://github.com/cosmos/cosmos-sdk/blob/0cbb37ff034396e837dc1cd81b82032cd068159e/store/prefixstore.go#L156

The basic pattern is: iterate over a range that definitely contains what we want, then trim it. I don't think we can get around that. My argument was given that we have to skip anyway we might as well choose `0x22` as it's simpler and is consistent with the logic used in cpIncr (i.e. `sdk.PrefixEndBytes` doesn't pad). Meh, who cares I guess.

Regarding inclusive endpoints - I didn't mean to make it inconsistent with Iterator, I think what I had in my head was it would be nice if the underlying iterator had an option for inclusive/exclusive endpoints - then these up layers would be less special-casey (I think the underlying DBs provide options for that IIRC). Though what I actually said doesn't make sense.